File: cpp/problems/leetcode/0141_linked_list_cycle/solution.cpp
// LeetCode 141
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) return true;
        }
        return false;
    }
};

int main() {
    ListNode n1(3), n2(2), n3(0), n4(-4);
    n1.next = &n2;
    n2.next = &n3;
    n3.next = &n4;
    n4.next = &n2;
    assert(Solution().hasCycle(&n1));
    ListNode a(1), b(2);
    a.next = &b;
    assert(!Solution().hasCycle(&a));
    cout << "0141 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0141_linked_list_cycle/solution.py
"""LeetCode 141: Linked List Cycle."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next  # type: ignore[union-attr]
            fast = fast.next.next
            if slow is fast:
                return True
        return False


if __name__ == "__main__":
    n1, n2, n3, n4 = ListNode(3), ListNode(2), ListNode(0), ListNode(-4)
    n1.next, n2.next, n3.next, n4.next = n2, n3, n4, n2
    assert Solution().hasCycle(n1)
    a = ListNode(1, ListNode(2))
    assert not Solution().hasCycle(a)
    print("0141 OK")
